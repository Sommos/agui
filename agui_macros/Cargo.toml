[package]
name = "agui_macros"
version = "0.2.0"
edition = "2021"
resolver = "2"
description = "The macro utilities of agui"
homepage = "https://github.com/Stumblinbear/agui"
repository = "https://github.com/Stumblinbear/agui"
keywords = [ ]
categories = [ ]
license = "MIT OR Apache-2.0"

[lib]
proc-macro = true

[features]
default = []
internal = []

[dependencies]
syn = { version = "1.0", features = ["full"] }
quote = "1.0"
darling = "0.13"
proc-macro-error = { version = "1", default-features = false }

[dev-dependencies]
trybuild = "1.0"

agui_core = { path = "../agui_core" }
agui_primitives = { path = "../agui_primitives" }