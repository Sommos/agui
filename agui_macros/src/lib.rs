extern crate proc_macro;

// use functional_widget::parse_functional_widget;
use proc_macro::TokenStream;
use widget_derive::parse_widget_derive;

mod builder;
// mod functional_widget;
mod widget_derive;

// #[proc_macro_attribute]
// pub fn functional_widget(args: TokenStream, input: TokenStream) -> TokenStream {
//     parse_functional_widget(args, input)
// }

#[proc_macro_derive(Widget, attributes(widget))]
pub fn widget_derive(input: TokenStream) -> TokenStream {
    parse_widget_derive(input)
}

#[proc_macro]
pub fn build(input: TokenStream) -> TokenStream {
    builder::build_impl(input.into()).into()
}
