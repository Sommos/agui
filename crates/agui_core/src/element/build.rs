use std::any::Any;

use crate::{callback::CallbackId, widget::Widget};

use super::{widget::ElementWidget, ElementBuildContext, ElementCallbackContext};

pub trait ElementBuild: ElementWidget {
    fn build(&mut self, ctx: ElementBuildContext) -> Widget;

    #[allow(unused_variables)]
    fn call(
        &mut self,
        ctx: ElementCallbackContext,
        callback_id: CallbackId,
        arg: Box<dyn Any>,
    ) -> bool;
}

impl std::fmt::Debug for Box<dyn ElementBuild> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct(self.widget_name()).finish_non_exhaustive()
    }
}
