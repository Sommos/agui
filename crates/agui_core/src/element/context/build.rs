use rustc_hash::FxHashSet;

use crate::{
    callback::CallbackQueue,
    element::{Element, ElementId},
    plugin::{
        context::{ContextPlugins, ContextPluginsMut},
        Plugins,
    },
    util::tree::Tree,
};

use super::{ContextElement, ContextMarkDirty};

pub struct ElementBuildContext<'ctx> {
    pub(crate) plugins: &'ctx mut Plugins,

    pub(crate) element_tree: &'ctx Tree<ElementId, Element>,
    pub(crate) dirty: &'ctx mut FxHashSet<ElementId>,

    pub(crate) element_id: ElementId,

    pub(crate) callback_queue: &'ctx CallbackQueue,
}

impl<'ctx> ContextPlugins<'ctx> for ElementBuildContext<'ctx> {
    fn get_plugins(&self) -> &Plugins {
        self.plugins
    }
}

impl<'ctx> ContextPluginsMut<'ctx> for ElementBuildContext<'ctx> {
    fn get_plugins_mut(&mut self) -> &mut Plugins {
        self.plugins
    }
}

impl ContextElement for ElementBuildContext<'_> {
    fn get_elements(&self) -> &Tree<ElementId, Element> {
        self.element_tree
    }

    fn get_element_id(&self) -> ElementId {
        self.element_id
    }
}

impl ContextMarkDirty for ElementBuildContext<'_> {
    fn mark_dirty(&mut self, element_id: ElementId) {
        self.dirty.insert(element_id);
    }
}
