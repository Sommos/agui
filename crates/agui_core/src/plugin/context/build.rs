use rustc_hash::FxHashSet;

use crate::{
    callback::CallbackQueue,
    element::{ContextElement, ContextMarkDirty, Element, ElementId},
    util::tree::Tree,
};

pub struct PluginBuildContext<'ctx> {
    pub(crate) element_tree: &'ctx Tree<ElementId, Element>,
    pub(crate) dirty: &'ctx mut FxHashSet<ElementId>,

    pub(crate) element_id: ElementId,

    pub(crate) callback_queue: &'ctx CallbackQueue,
}

impl ContextElement for PluginBuildContext<'_> {
    fn get_elements(&self) -> &Tree<ElementId, Element> {
        self.element_tree
    }

    fn get_element_id(&self) -> ElementId {
        self.element_id
    }
}

impl ContextMarkDirty for PluginBuildContext<'_> {
    fn mark_dirty(&mut self, element_id: ElementId) {
        self.dirty.insert(element_id);
    }
}
