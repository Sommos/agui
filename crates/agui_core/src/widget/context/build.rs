use std::{any::TypeId, ops::Deref};

use fnv::{FnvHashMap, FnvHashSet};

use crate::{
    callback::{Callback, CallbackContext, CallbackFn, CallbackFunc, CallbackId, CallbackQueue},
    context::ContextMut,
    element::{Element, ElementId},
    inheritance::Inheritance,
    unit::{Data, Key},
    util::tree::Tree,
    widget::{InheritedWidget, Widget, WidgetKey, WidgetRef, WidgetState, WidgetView},
};

use super::{ContextStatefulWidget, ContextWidget, ContextWidgetMut};

pub struct BuildContext<'ctx, W>
where
    W: Widget,
{
    pub(crate) element_tree: &'ctx mut Tree<ElementId, Element>,
    pub(crate) dirty: &'ctx mut FnvHashSet<ElementId>,
    pub(crate) callback_queue: &'ctx CallbackQueue,

    pub(crate) element_id: ElementId,

    pub(crate) inheritance: &'ctx mut Inheritance,

    pub widget: &'ctx W,
    pub state: &'ctx mut W::State,

    pub(crate) callbacks: FnvHashMap<CallbackId, Box<dyn CallbackFunc<W>>>,

    pub(crate) keyed_children: FnvHashSet<Key>,
}

impl<W> Deref for BuildContext<'_, W>
where
    W: Widget,
{
    type Target = W;

    fn deref(&self) -> &Self::Target {
        self.widget
    }
}

impl<W> ContextMut for BuildContext<'_, W>
where
    W: Widget,
{
    fn mark_dirty(&mut self, element_id: ElementId) {
        self.dirty.insert(element_id);
    }

    fn call<A>(&mut self, callback: &Callback<A>, arg: A)
    where
        A: Data,
    {
        self.callback_queue.call(callback, arg);
    }

    fn call_unchecked(&mut self, callback_id: CallbackId, arg: Box<dyn Data>) {
        self.callback_queue.call_unchecked(callback_id, arg);
    }

    fn call_many<A>(&mut self, callbacks: &[Callback<A>], arg: A)
    where
        A: Data,
    {
        self.callback_queue.call_many(callbacks, arg);
    }

    fn call_many_unchecked(&mut self, callback_ids: &[CallbackId], arg: Box<dyn Data>) {
        self.callback_queue.call_many_unchecked(callback_ids, arg);
    }
}

impl<W> ContextWidget for BuildContext<'_, W>
where
    W: Widget,
{
    type Widget = W;

    fn get_elements(&self) -> &Tree<ElementId, Element> {
        self.element_tree
    }

    fn get_element_id(&self) -> ElementId {
        self.element_id
    }

    fn get_widget(&self) -> &W {
        self.widget
    }
}

impl<W> ContextStatefulWidget for BuildContext<'_, W>
where
    W: Widget,
{
    fn get_state(&self) -> &W::State {
        self.state
    }

    fn get_state_mut(&mut self) -> &mut W::State {
        self.state
    }

    fn set_state<F>(&mut self, func: F)
    where
        F: FnOnce(&mut W::State),
    {
        func(self.state);
    }
}

impl<W> ContextWidgetMut for BuildContext<'_, W>
where
    W: Widget,
{
    fn depend_on_inherited_widget<I>(&mut self) -> Option<&mut I::State>
    where
        I: InheritedWidget + 'static,
    {
        let type_id = TypeId::of::<I>();

        // self.inheritance.listening_to.insert(type_id);

        // if let Some(inheritance_scope_id) = self.inheritance.scope {
        //     // Grab our ancestor inherited node from the tree. This contains the mappings of available
        //     // types that we can consume.
        //     let target_inherited_element_id = self
        //         .element_tree
        //         .get(inheritance_scope_id)
        //         .expect("ancestor inherited node does not exist")
        //         .inheritance_scope
        //         .available
        //         .get(&type_id)
        //         .copied();

        //     if let Some(target_inherited_element_id) = target_inherited_element_id {
        //         let target_inherited_element = self
        //             .element_tree
        //             .get_mut(target_inherited_element_id)
        //             .expect("inherited element does not exist");

        //         let inheritance_node = &mut target_inherited_element.inheritance_scope;

        //         // Add this widget as a listener on the element we're inheriting from
        //         inheritance_node.listeners.insert(self.element_id);

        //         self.inheritance
        //             .depends_on
        //             .insert(target_inherited_element_id);

        //         let instance = target_inherited_element
        //             .downcast_mut::<WidgetInstance<I>>()
        //             .expect("inherited widget downcast failed");

        //         return Some(&mut instance.state);
        //     }
        // }

        None
    }

    fn callback<A, F>(&mut self, func: F) -> Callback<A>
    where
        A: Data,
        F: Fn(&mut CallbackContext<W>, &A) + 'static,
    {
        let callback = Callback::new::<F, W>(self.element_id, self.callback_queue.clone());

        self.callbacks
            .insert(callback.get_id().unwrap(), Box::new(CallbackFn::new(func)));

        callback
    }
}

impl<W> BuildContext<'_, W>
where
    W: WidgetView + WidgetState,
{
    pub fn key<C>(&mut self, key: Key, widget: C) -> WidgetRef
    where
        C: Widget,
    {
        if self.keyed_children.contains(&key) {
            panic!("cannot use the same key twice in a widget");
        }

        self.keyed_children.insert(key);

        WidgetRef::new_with_key(
            Some(match key {
                Key::Local(_) => WidgetKey(Some(self.element_id), key),
                Key::Global(_) => WidgetKey(None, key),
            }),
            widget,
        )
    }
}
