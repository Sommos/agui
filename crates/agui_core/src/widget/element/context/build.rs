use rustc_hash::FxHashSet;

use crate::{
    callback::CallbackQueue,
    element::{Element, ElementId},
    plugin::{
        context::{ContextPlugins, ContextPluginsMut},
        Plugins,
    },
    util::tree::Tree,
    widget::ContextWidget,
};

pub struct WidgetBuildContext<'ctx> {
    pub(crate) plugins: Plugins<'ctx>,

    pub(crate) element_tree: &'ctx Tree<ElementId, Element>,

    pub(crate) dirty: &'ctx mut FxHashSet<ElementId>,
    pub(crate) callback_queue: &'ctx CallbackQueue,

    pub(crate) element_id: ElementId,
}

impl ContextWidget for WidgetBuildContext<'_> {
    fn get_elements(&self) -> &Tree<ElementId, Element> {
        self.element_tree
    }

    fn get_element_id(&self) -> ElementId {
        self.element_id
    }
}

impl<'ctx> ContextPlugins<'ctx> for WidgetBuildContext<'ctx> {
    fn get_plugins(&self) -> &Plugins<'ctx> {
        &self.plugins
    }
}

impl<'ctx> ContextPluginsMut<'ctx> for WidgetBuildContext<'ctx> {
    fn get_plugins_mut(&mut self) -> &mut Plugins<'ctx> {
        &mut self.plugins
    }
}

impl WidgetBuildContext<'_> {
    pub fn mark_dirty(&mut self, element_id: ElementId) {
        self.dirty.insert(element_id);
    }
}
