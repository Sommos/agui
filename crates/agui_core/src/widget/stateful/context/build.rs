use std::{any::TypeId, marker::PhantomData};

use fnv::{FnvHashMap, FnvHashSet};

use crate::{
    callback::{Callback, CallbackId, CallbackQueue},
    element::{Element, ElementId},
    unit::{Data, Key},
    util::tree::Tree,
    widget::{
        AnyWidget, ContextInheritedMut, ContextWidget, Inheritance, InheritedWidget, WidgetKey,
        WidgetRef, WidgetState,
    },
};

use super::{ContextWidgetState, StatefulCallbackContext};

pub trait StatefulCallbackFunc<W> {
    #[allow(clippy::borrowed_box)]
    fn call(&self, ctx: &mut StatefulCallbackContext<W>, args: &Box<dyn Data>);
}

pub struct StatefulCallbackFn<W, A, F>
where
    A: 'static,
    F: Fn(&mut StatefulCallbackContext<W>, &A),
{
    phantom: PhantomData<(W, A, F)>,

    func: F,
}

impl<W, A, F> StatefulCallbackFn<W, A, F>
where
    A: 'static,
    F: Fn(&mut StatefulCallbackContext<W>, &A),
{
    pub fn new(func: F) -> Self {
        Self {
            phantom: PhantomData,

            func,
        }
    }
}

impl<W, A, F> StatefulCallbackFunc<W> for StatefulCallbackFn<W, A, F>
where
    A: Data,
    F: Fn(&mut StatefulCallbackContext<W>, &A),
{
    fn call(&self, ctx: &mut StatefulCallbackContext<W>, args: &Box<dyn Data>) {
        let args = args
            .downcast_ref::<A>()
            .expect("failed to downcast callback args");

        (self.func)(ctx, args)
    }
}

pub struct StatefulContext<'ctx, S>
where
    S: WidgetState,
{
    pub(crate) element_tree: &'ctx Tree<ElementId, Element>,
    pub(crate) dirty: &'ctx mut FnvHashSet<ElementId>,
    pub(crate) callback_queue: &'ctx CallbackQueue,

    pub(crate) element_id: ElementId,

    pub(crate) callbacks: &'ctx mut FnvHashMap<CallbackId, Box<dyn StatefulCallbackFunc<S>>>,

    pub(crate) inheritance: &'ctx mut Inheritance,

    pub(crate) keyed_children: FnvHashSet<Key>,

    pub widget: &'ctx S::Widget,
    pub(crate) state: &'ctx S,
}

impl<S> ContextWidget<S> for StatefulContext<'_, S>
where
    S: WidgetState,
{
    fn get_elements(&self) -> &Tree<ElementId, Element> {
        self.element_tree
    }

    fn get_element_id(&self) -> ElementId {
        self.element_id
    }
}

impl<S> ContextWidgetState<S> for StatefulContext<'_, S>
where
    S: WidgetState,
{
    fn get_state(&self) -> &S {
        self.state
    }
}

impl<'ctx, S: 'static> StatefulContext<'ctx, S>
where
    S: WidgetState,
{
    pub fn get_widget(&self) -> &S::Widget {
        self.widget
    }

    pub fn mark_dirty(&mut self, element_id: ElementId) {
        self.dirty.insert(element_id);
    }

    pub fn key<C>(&mut self, key: Key, widget: C) -> WidgetRef
    where
        C: AnyWidget,
    {
        if self.keyed_children.contains(&key) {
            panic!("cannot use the same key twice in a widget");
        }

        self.keyed_children.insert(key);

        WidgetRef::new_with_key(
            Some(match key {
                Key::Local(_) => WidgetKey(Some(self.element_id), key),
                Key::Global(_) => WidgetKey(None, key),
            }),
            widget,
        )
    }

    pub fn callback<A, F>(&mut self, func: F) -> Callback<A>
    where
        A: Data,
        F: Fn(&mut StatefulCallbackContext<S>, &A) + 'static,
    {
        let callback = Callback::new::<F, S>(self.element_id, self.callback_queue.clone());

        self.callbacks.insert(
            callback.get_id().unwrap(),
            Box::new(StatefulCallbackFn::new(func)),
        );

        callback
    }
}

impl<S> ContextInheritedMut for StatefulContext<'_, S>
where
    S: WidgetState,
{
    fn depend_on_inherited_widget<I>(&mut self) -> Option<&mut I>
    where
        I: InheritedWidget + 'static,
    {
        let type_id = TypeId::of::<I>();

        // self.inheritance.listening_to.insert(type_id);

        // if let Some(inheritance_scope_id) = self.inheritance.scope {
        //     // Grab our ancestor inherited node from the tree. This contains the mappings of available
        //     // types that we can consume.
        //     let target_inherited_element_id = self
        //         .element_tree
        //         .get(inheritance_scope_id)
        //         .expect("ancestor inherited node does not exist")
        //         .inheritance_scope
        //         .available
        //         .get(&type_id)
        //         .copied();

        //     if let Some(target_inherited_element_id) = target_inherited_element_id {
        //         let target_inherited_element = self
        //             .element_tree
        //             .get_mut(target_inherited_element_id)
        //             .expect("inherited element does not exist");

        //         let inheritance_node = &mut target_inherited_element.inheritance_scope;

        //         // Add this widget as a listener on the element we're inheriting from
        //         inheritance_node.listeners.insert(self.element_id);

        //         self.inheritance
        //             .depends_on
        //             .insert(target_inherited_element_id);

        //         let instance = target_inherited_element
        //             .downcast_mut::<WidgetInstance<I>>()
        //             .expect("inherited widget downcast failed");

        //         return Some(&mut instance.state);
        //     }
        // }

        None
    }
}
