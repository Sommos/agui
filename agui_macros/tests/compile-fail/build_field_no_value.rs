use agui_core::WidgetRef;
use agui_macros::build;
use agui_primitives::Quad;

fn main() {
    let _widget: WidgetRef = build! {
        Quad {
            color:
        }
    };
}
