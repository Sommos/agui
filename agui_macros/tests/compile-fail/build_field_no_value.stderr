error: unexpected end of input, expected expression
 --> tests/compile-fail/build_field_no_value.rs:9:9
  |
9 |         }
  |         ^

warning: unused import: `agui_primitives::Quad`
 --> tests/compile-fail/build_field_no_value.rs:3:5
  |
3 | use agui_primitives::Quad;
  |     ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
